jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/ResourceService.java

```diff
@@ -469,66 +466,130 @@ protected void sendWelcome(HttpContent content, String pathInContext, boolean en
     }
 
-    public enum WelcomeActionType
+    /**
+     * <p>How welcome targets should be processed.</p>
+     */
+    public enum WelcomeMode
     {
+        /**
+         * The welcome target is used in an HTTP redirect
+         * as the value of the {@code Location} header.
+         */
         REDIRECT,
-        SERVE
+        /**
+         * The welcome target is served either as a real file
+         * or forwarded to be processed again.
+         */
+        SERVE,
+        /**
+         * The welcome target is wrapped in a new request
+         * that is re-handled.
+         */
+        REHANDLE
     }
 
     /**
-     * Behavior for a potential welcome action
-     * as determined by {@link ResourceService#processWelcome(Request, Response)}
-     *
-     * <p>
-     * For {@link WelcomeActionType#REDIRECT} this is the resulting `Location` response header.
-     * For {@link WelcomeActionType#SERVE} this is the resulting path to for welcome serve, note that
-     * this is just a path, and can point to a real file, or a dynamic handler for
-     * welcome processing (such as Jetty core Handler, or EE Servlet), it's up
-     * to the implementation of {@link ResourceService#welcome(Request, Response, Callback)}
-     * to handle the various action types.
-     * </p>
+     * <p>A welcome target paired with how to process it.</p>
      *
-     * @param type the type of action
-     * @param target The target URI path of the action.
+     * @param target the welcome target
+     * @param mode the welcome mode
      */
-    public record WelcomeAction(WelcomeActionType type, String target) {}
+    public record WelcomeAction(String target, WelcomeMode mode)
+    {
+    }
 
-    private boolean welcome(Request request, Response response, Callback callback) throws IOException
+    private boolean welcome(Request request, Response response, Callback callback) throws Exception
     {
-        WelcomeAction welcomeAction = processWelcome(request, response);
+        WelcomeAction welcomeAction = processWelcome(request);
         if (LOG.isDebugEnabled())
-        {
-            LOG.debug("welcome(req={}, resp={}, callback={}) welcomeAction={}",
-                request, response, callback, welcomeAction);
-        }
+            LOG.debug("welcome(req={}, rsp={}, cbk={}) welcomeAction={}", request, response, callback, welcomeAction);
+
         if (welcomeAction == null)
             return false;
 
-        handleWelcomeAction(request, response, callback, welcomeAction);
-        return true;
+        return handleWelcomeAction(request, response, callback, welcomeAction);
     }
 
-    protected void handleWelcomeAction(Request request, Response response, Callback callback, WelcomeAction welcomeAction) throws IOException
+    protected boolean handleWelcomeAction(Request request, Response response, Callback callback, WelcomeAction welcomeAction) throws Exception
```

Comment:
The return value is used in `sendWelcome()`: if `false` the directory is served.

The return value may be `false` in case of re-handling, where the welcome URI goes again through the handlers, and none handles it.

Reply:
I understand how the boolean return is used.  What is unclear is the two different meanings of a false return from serveWelcome: 1) No welcome file found; 2) a welcome file was found but could not be served.

If a welcome file is found but cannot be served (for some unknown reason), is it correct that a directory is served?   I think perhaps not.  Maybe a 500 would be more appropriate?

Reply:
It has the same semantic of `Handler.handle()`, as specified in the javadocs of `serveWelcome()`.